install/lnst-ctl.1.in
.TH LNST-CTL "1" "@date@"

.\" To view this file while editing, run it through groff:
.\"   groff -Tascii -man lnst-ctl.1 | less -r

.\" This file is a template that contains placeholders for
.\" certain values. It needs to be processed before it can
.\" be used as a man-page.

.SH NAME
lnst-ctl \- Linux Network Stack Test controller
.SH SYNOPSIS
.BR
.B lnst-ctl
[\fIOPTION\fR]... \fIACTION\fR \fIRECIPE\fR...
.SH DESCRIPTION
Linux Network Stack Test is a tool that supports development and execution
of automated and portable network tests. For detailed description of the
architecture of LNST please refer to project website
(link listed on
.B INTERNET RESOURCES
bellow).

.B lnst-ctl
is a script used for managing slaves. It reads a recipe XML file as an
input and executes tests on the specified slave machines.
.SH OPTIONS
.TP
.BI "\-A, \-\-override\-alias"\ name=value
Define top\-level alias that will override any other definitions in the recipe
.TP
.BI "\-a, \-\-define\-alias"\ name=value
Define top\-level alias
.TP
.BI "\-c, \-\-config" =file
Load a custom configuration file in addition to
.I /etc/lnst-ctl.conf
and
.I ~/lnst/lnst-ctl.conf
that are loaded by default.
.TP
.B \-d, \-\-debug
Toggle emitting debugging messages.
.TP
.B \-\-dump\-config
Dumps the join of all loaded configuration files on stdout and exits
.TP
.B \-h, \-\-help
Display usage of this command.
.TP
.B \-m, \-\-no\-colours
Disable coloured terminal output.
.TP
.B \-o, \-\-disable\-pool\-checks
Skip the pool machine availability checks.
.TP
.B \-p, \-\-packet\-capture
Capture and log all ongoing network communication between the slaves
during the test.
.TP
.BI "\-\-pools" =name[,...]
Restricts which pools to use for matching, value is a comma separated list of values
.TP
.BI "\-\-pools=" path
Restrict which pool to use for matching, value is single path to a pool directory
.TP
.B "\-r, \-\-reduce-sync"
Disable synchronization of module and tool resources to slave machines,
when python tasks are used in the recipe. Instead the user can manually
synchronize resources from the python task itself. This option has no effect if
the recipe only contains xml tasks.
.TP
.BI "\-s, \-\-xslt-url" =URL
URL to a XSLT document that will be used when transforming the xml result file,
only useful when -t is used as well
.TP
.BI "\-t, \-\-html" =file
Generate a formatted result html
.TP
.B \-u, \-\-multi\-match
Run each recipe with every pool match possible
.TP
.BI "\-x, \-\-result" =file
Write results in XML format to the specified
.IR file .
.TP
.B \-v, \-\-verbose
Extends output of list_pools action
.PP
.I RECIPE
can be either a LNST recipe file or a directory containing recipe files.
One or more files/directories can be passed at the same time. Each recipe
is a XML file. The format is described in detail in the online documentation
available at LNST wiki (link is listed on
.B INTERNET RESOURCES
bellow).

.I ACTION
can be one of
.BR "run",
.BR "config_only",
.BR "deconfigure" ",
.BR "match_setup",\ and
.BR "list_pools".

When it is set to
.BR "run",
all provided recipe files will be sequentially executed. The network will
be configured accordingly to each recipe file and clean up
afterwards.

In case of
.BR "config_only",
the recipes provided are processed up to the point where all the machines
are prepared and properly configured. No tests will be executed and the
cleanup phase is omitted as well. This mode of operation is useful if you
want to perform some specific actions or tests outside of LNST using
the configuration specified in a recipe file.

In case of
.BR "deconfigure",
the configuration from a previous config_only run will be removed, including
dynamically created devices on virtual machines. There is no need to specify a
recipe for this action.

In case of
.BR "list_pools",
lnst-ctl looks up all the pools entered in configs and dumps info about them.
When used without
.B \-v
or
.B \-\-verbose
option, only pool names and paths are printed.
With verbose option, all viable slave machine configs are printed in more
human readable format.

At last, when the
.I ACTION
is set to
.BR "match_setup",
.B lnst-ctl
will only try to match the configuration in the recipe with its local
pools and print the resulting match, in case it was successful or an
appropriate error message, in case it was not.

.SH CONFIGURATION
.B lnst-ctl
will look for configuration in the following locations:

.I @conf_dir@lnst-ctl.conf
.br
.I ~/.lnst/lnst-ctl.conf

For further description of file format or configuration options available,
please refer to the configuration file itself, where each option is
documented.

.SH EXIT STATUS
.B lnst-ctl
will terminate with one of the following statuses:

    0    All tests passed, no errors encountered
    1    At least one test failed
    2    There was an error during the execution

.SH INTERNET RESOURCES
Project Homepage:      http://lnst-project.org/
.br
Online Documentation:  https://github.com/jpirko/lnst/wiki#learn
.br
Git Source Tree:       https://github.com/jpirko/lnst
.br
Mailing List:          <lnst-developers@lists.fedorahosted.org>

.SH AUTHORS
\fBJiri Pirko\fR <jiri@resnulli.us>
.br
\fBJan Tluka\fR <jtluka@redhat.com>
.br
\fBOndrej Lichtner\fR <olichtne@redhat.com>
.br
\fBJiri Prochazka\fR <jprochaz@redhat.com>
.br
\fBJiri Zupka\fR <jzupka@redhat.com>
.br
\fBRadek Pazdera\fR <radek@pazdera.co.uk>
.SH COPYRIGHT
Copyright (C) 2011\-2015 Red Hat, Inc.

LNST is distributed under the GNU General Public License version 2. See
the file "COPYING" in the source distribution for information on terms &
conditions for accessing and otherwise using LNST.


install/lnst-ctl.bash
#!/bin/bash

# Bash completion script for lnst-ctl command
# Author: Radek Pazdera <rpazdera@redhat.com>

_list_has_item()
{
    for entry in $1; do
        [ "$entry" == "$2" ] && return 0
    done

    return 1
}

_lnst_ctl()
{
    local SHORT_OPTS="-a -A -c -d -h -m -o -p -x -v"
    local LONG_OPTS="--define-alias --override-alias --config --debug \
                     --help --no-colours --disable-pool-checks \
                     --packet-capture --result --verbose"
    local REQUIRE_ARG="-a --define-alias \
                       -A --override-alias \
                       -c --config \
                       -x --result"
    local ACTIONS="config_only match_setup run list_pools"

    local cur=${COMP_WORDS[COMP_CWORD]}
    local prev=${COMP_WORDS[COMP_CWORD-1]}

    # Look for option arguments first
    case "$prev" in
        -a|--define-alias) return 0 ;;
        -A|--override-alias) return 0 ;;
        -c|--config)
            _filedir
            return 0
            ;;
        -x|--result)
            _filedir
            return 0
            ;;
    esac

    # Complete long and shor options
    if [[ "$cur" == --* ]]; then
        COMPREPLY=( $(compgen -W "$LONG_OPTS" -- $cur) )
        [ ${#COMPREPLY[@]} = 1 ] && COMPREPLY=$(printf %q%s "$COMPREPLY" " ")
        return 0
    elif [[ "$cur" == -* ]]; then
        COMPREPLY=( $(compgen -W "$SHORT_OPTS" -- $cur) )
        [ ${#COMPREPLY[@]} = 1 ] && COMPREPLY=$(printf %q%s "$COMPREPLY" " ")
        return 0
    fi

    # Check if the action positional argument has
    # already been specified somewhere
    for (( n=1; n < $COMP_CWORD; n++ )); do
        local word=${COMP_WORDS[n]}
        local prev_word=${COMP_WORDS[n-1]}

        # Is it an option?
        if [[ "$word" == --* ]] || [[ "$word" == -* ]]; then
            continue
        else
            # Is the previous word an option that requires and argument?
            _list_has_item "$REQUIRE_ARG" "$prev_word"
            if [ $? -eq 0 ]; then
                continue
            else
                # Is the positional argument an action?
                _list_has_item "$ACTIONS" "$word"
                if [ $? -eq 0 ]; then
                    # Action positional argument was found.
                    # Therefore, we will suggest files.
                    _filedir
                    return 0
                fi
            fi
        fi
    done

    # No action defined yet, we will suggest actions.
    COMPREPLY=( $(compgen -W "$ACTIONS" -- $cur) )
    [ ${#COMPREPLY[@]} = 1 ] && COMPREPLY=$(printf %q%s "$COMPREPLY" " ")

    return 0
}

complete -o nospace -F _lnst_ctl lnst-ctl
